app/scripts/controllers/network/network-controller.js

```diff
@@ -179,62 +197,106 @@ export default class NetworkController extends EventEmitter {
       );
-      this._setNetworkState('loading');
-      this._clearNetworkDetails();
+      this._resetNetworkId();
+      this._resetNetworkStatus();
+      this._resetNetworkDetails();
       return;
     }
 
-    // Ping the RPC endpoint so we can confirm that it works
-    const initialNetwork = this.networkStore.getState();
-    const { type } = this.providerStore.getState();
-    const isInfura = INFURA_PROVIDER_TYPES.includes(type);
-
-    if (isInfura) {
-      this._checkInfuraAvailability(type);
-    } else {
-      this.emit(NETWORK_EVENTS.INFURA_IS_UNBLOCKED);
-    }
+    this.once(NETWORK_EVENTS.NETWORK_DID_CHANGE, () => {
+      networkChanged = true;
+    });
 
-    let networkVersion;
-    let networkVersionError;
     try {
-      networkVersion = await this._getNetworkId();
+      const results = await Promise.all([
+        this._getNetworkId(),
+        this._determineEIP1559Compatibility(),
+      ]);
+      networkId = results[0];
+      supportsEIP1559 = results[1];
+      networkStatus = NetworkStatus.Available;
     } catch (error) {
-      networkVersionError = error;
+      let data;
+      if (hasProperty(error, 'data')) {
+        data = error.data;
+      } else if (hasProperty(error, 'code')) {
+        try {
+          data = JSON.parse(error.message);
+        } catch {
+          // error.message must not be JSON for some reason
+        }
+      }
+      if (data) {
+        if (data.error === INFURA_BLOCKED_KEY) {
+          networkStatus = NetworkStatus.Blocked;
+        } else {
+          networkStatus = NetworkStatus.Unavailable;
+        }
+      } else {
+        log.warn(
+          'NetworkController - could not determine network status',
+          error,
+        );
+        networkStatus = NetworkStatus.Unknown;
+      }
     }
-    if (initialNetwork !== this.networkStore.getState()) {
+
+    if (networkChanged) {
+      // If the network has changed, then `lookupNetwork` either has been or is
+      // in the process of being called, so we don't need to go further.
       return;
     }
 
-    if (networkVersionError) {
-      this._setNetworkState('loading');
-      // keep network details in sync with network state
-      this._clearNetworkDetails();
+    this.networkStatusStore.putState(networkStatus);
+
+    if (networkStatus === NetworkStatus.Available) {
+      this.networkIdStore.putState(networkId);
+      this.networkDetails.updateState({
+        EIPS: {
+          ...this.networkDetails.getState().EIPS,
+          1559: supportsEIP1559,
+        },
+      });
+      this.emit(NETWORK_EVENTS.INFURA_IS_UNBLOCKED);
     } else {
-      this._setNetworkState(networkVersion);
-      // look up EIP-1559 support
-      await this.getEIP1559Compatibility();
+      this._resetNetworkId();
+      this._resetNetworkDetails();
+      if (networkStatus === NetworkStatus.Blocked) {
+        this.emit(NETWORK_EVENTS.INFURA_IS_BLOCKED);
+      }
```

Comment:
The current `lookupNetwork` function also emits this for non-Infura networks before even performing any lookup. That ensures this doesn't get stuck enabled on non-Infura networks.

We could handle that same case here in an `else` block. Or we could do something different, like check what type of network this is earlier on (like in the current `lookupNetwork`, or emit this when the user switches to a custom RPC endpoint instead.

Reply:
> That ensures this doesn't get stuck enabled on non-Infura networks.

Would this happen if the user switches to an Infura network, receives a blocked error, then switches to a network that is unavailable? As you allude to, maybe we always emit `infuraIsUnblocked` when switching to a non-Infura network, regardless of network status?

Reply:
>Would this happen if the user switches to an Infura network, receives a blocked error, then switches to a network that is unavailable?

Yes, exactly

>maybe we always emit `infuraIsUnblocked` when switching to a non-Infura network, regardless of network status?

Sure, that would make sense to me

Reply:
Okay, this is done in [5cad8ea](https://github.com/MetaMask/metamask-extension/pull/17556/commits/5cad8eacc7f4cf6565cc74e01b0f4208d89455c6).